### PR TITLE
Features/staging install script

### PR DESCRIPTION
**Pull request by ifrit98**
* Add a nice script to install and run on staging to minimize friction for developer UX experience.
* Depends on subtensor branch: `localnet/specify-init-wallet-bal` https://github.com/opentensor/subtensor/pull/210 (This is now merged into main!) <- merged!

### CODE REVIEW

File: scripts/install_staging.sh [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+# Install rust and cargo
+curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
+
+# Update your shell's source to include Cargo's path
+source "$HOME/.cargo/env"
+
+# Clone subtensor and enter the directory
+cd ../
+git clone https://github.com/opentensor/subtensor.git
```
Comment (steffencruz) [inline scripts/install_staging.sh:15]: maybe check if subtensor already exists. also make sure that when the demo ends the subtensor setup is back to its original state

File: scripts/install_staging.sh [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+
+# Navigate to your project directory
+cd ../bittensor-subnet-template
+
+# Install the bittensor-subnet-template python package
+python -m pip install -e .
+
+# Create a coldkey for the owner role
+btcli wallet new_coldkey --wallet.name owner2
```
Comment (steffencruz) [inline scripts/install_staging.sh:39]: we can split the file into multiple parts:
- `install.sh`: takes care of dependencies and installation
- `create.sh`: creates wallets, registers subnet, everything up until `btcli wallet overview`
- `demo.sh`: runs demo (tmux)

File: scripts/install_staging.sh [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+# Update to the nightly version of rust
+./scripts/init.sh
+
+# Initialize your local subtensor chain in development mode. This command will set up and run a local subtensor network.
+# Run this in the background (or send to tmux/forked shell?)
+
+# Check if inside a tmux session
+# Start a new tmux session and create a new pane, but do not switch to it
+tmux new-session -d -s localnet -n 'script' 'bash ./scripts/localnet.sh'
```
Comment (steffencruz) [inline scripts/install_staging.sh:26]: need to install tmux

File: scripts/install_staging.sh [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+./scripts/init.sh
+
+# Navigate to your project directory
+cd ../bittensor-subnet-template
+
+# Install the bittensor-subnet-template python package
+python -m pip install -e .
+
+# Create a coldkey for the owner role
+btcli wallet new_coldkey --wallet.name owner --no_password --no_prompt
```
Comment (steffencruz) [inline scripts/install_staging.sh:61]: Make configurable 
```suggestion
wallet=${1:-owner}

btcli wallet new_coldkey --wallet.name $wallet --no_password --no_prompt
```

File: scripts/install_staging.sh [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+# Detect OS and call the appropriate function
+if [[ "$OSTYPE" == "darwin"* ]]; then
+    install_mac
+elif [[ "$OSTYPE" == "linux-gnu"* ]]; then
+    install_ubuntu
+else
+    echo "Unsupported operating system."
+fi
+
```
Comment (steffencruz) [inline scripts/install_staging.sh:36]: We should thinkg carefully about the following:
1. First time running the script 
2. All additional times (reinstall / create wallet 

I think it's a better design to have two scripts (or at least logical sections that only run when necessary):
1. Build/install (OS requirements, libraries, btcli interactions)
2. Test/run (only python scripts)

Comment (ifrit98) [reply scripts/install_staging.sh:36]: How about this?

```bash
#!/bin/bash

# Section 1: Build/Install
# This section is for first-time setup and installations.

install_dependencies() {
    # Function to install packages on macOS
    install_mac() {
        which brew > /dev/null
        if [ $? -ne 0 ]; then
            echo "Installing Homebrew..."
            /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
        fi
        echo "Updating Homebrew packages..."
        brew update
        echo "Installing required packages..."
        brew install make llvm curl libssl protobuf tmux
    }

    # Function to install packages on Ubuntu/Debian
    install_ubuntu() {
        echo "Updating system packages..."
        sudo apt update
        echo "Installing required packages..."
        sudo apt install --assume-yes make build-essential git clang curl libssl-dev llvm libudev-dev protobuf-compiler tmux
    }

    # Detect OS and call the appropriate function
    if [[ "$OSTYPE" == "darwin"* ]]; then
        install_mac
    elif [[ "$OSTYPE" == "linux-gnu"* ]]; then
        install_ubuntu
    else
        echo "Unsupported operating system."
        exit 1
    fi

    # Install rust and cargo
    curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh

    # Update your shell's source to include Cargo's path
    source "$HOME/.cargo/env"
}

# Call install_dependencies only if it's the first time running the script
if [ ! -f ".dependencies_installed" ]; then
    install_dependencies
    touch .dependencies_installed
fi

# Section 2: Test/Run
# This section is for running and testing the setup.

# Logic for setting up and running the environment
setup_environment() {
    # Clone subtensor and enter the directory
    if [ ! -d "subtensor" ]; then
        git clone https://github.com/opentensor/subtensor.git
    fi
    cd subtensor
    git pull

    # Update to the nightly version of rust
    ./scripts/init.sh

    cd ../bittensor-subnet-template

    # Install the bittensor-subnet-template python package
    python -m pip install -e .

    # Create and set up wallets
    # This section can be skipped if wallets are already set up
    if [ ! -f ".wallets_setup" ]; then
        btcli wallet new_coldkey --wallet.name owner --no_password --no_prompt
        btcli wallet new_coldkey --wallet.name miner --no_password --no_prompt
        btcli wallet new_hotkey --wallet.name miner --wallet.hotkey default --no_prompt
        btcli wallet new_coldkey --wallet.name validator --no_password --no_prompt
        btcli wallet new_hotkey --wallet.name validator --wallet.hotkey default --no_prompt
        touch .wallets_setup
    fi

    # Additional script logic goes here...
}

# Call setup_environment every time
setup_environment

# Additional logic for running and testing the setup goes here...

```

Comment (ifrit98) [reply scripts/install_staging.sh:36]: This way we can have a touch file to denote if we need to resetup wallets and whatnot.

**Review by Eugene-hu — APPROVED** [review]

**Pull request merged**